event/brokerPool.go

```diff
@@ -0,0 +1,130 @@
+package event
+
+import (
+	"context"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+)
+
+type EventSourceFactory func(ctx context.Context, id string, arg interface{}) (EventSource, error)
+
+// BrokerPool manages a group of brokers which can be accessed and retrieved via an ID.
+// It also simplifies the access to the brokers themselves, providing a single Subscribe
+// function which automatically makes the unsubscription when the context is cancelled.
+//
+// The EventSourceFactory passed as argument to the constructor should create a new
+// EventSource channel, using the optional factoryArg that is passed on subscription.
+//
+// In the simplest case, for example for messaging events between internal application
+// components, there could be a single set of EventSources and the factory would grab them
+// by their ID and return otherwise return an error for inexistent event source. In more
+// complex scenarios, these event sources could be actual subscriptions to some remote
+// messaging service for example Redis or RabbitMQ.
+type BrokerPool interface {
+	Subscribe(ctx context.Context, id string, factoryArg interface{}) (EventSource, error)
+}
+
+func NewPool(factory EventSourceFactory) BrokerPool {
+	return &brokerPool{
+		factory: factory,
+		brokers: map[string]*poolEntry{},
+	}
+}
+
+type brokerPool struct {
+	factory EventSourceFactory
+
+	lock    sync.RWMutex
+	brokers map[string]*poolEntry
+}
+
+type poolEntry struct {
+	id     string
+	broker Broker
+	stopFn func()
+}
+
+func (p *brokerPool) Subscribe(ctx context.Context, id string, factoryArg interface{}) (EventSource, error) {
+	p.lock.RLock()
+	if entry, ok := p.brokers[id]; ok {
+		defer p.lock.RUnlock()
+		return p.subscribeLocked(ctx, entry)
+	}
+	p.lock.RUnlock()
+
+	entry, err := p.createBroker(id, factoryArg)
```

Comment:
⛏️: have you thought about any kind of control here to prevent duplicate creation?

Reply:
I guess one way would be to grab the write `lock` before we started creating the broker, then we could check if someone else already created it before trying to create a new one. That would be sub-part though since we'd look the whole pool on every specific broker creation which calls an external factory function that we don't know if will be expensive.

One improvement I would like to make to all these implementations, is to start using a couple of the `sync` abstractions to avoid all this locking around here. For the broker initialization itself, an ideal would be to have a `sync.Map` which holds a struct that has a broker and a `sync.Once`, and we use the `sync.Once` to initialize the broker itself. I'll see if I can make it real quick here cause it sounds interesting 👀